### PR TITLE
[pinet2-p5] Pre-release review for pinet2

### PR DESCRIPTION
**Pull request by Roy-Kid**
Hi everyone,

We are moving to freeze the code and prepare to test the performance systemically on different datasets. Here, I start a new pull request for the newly designed architecture of Pinet2-p5. This pull request is our shared code in the future before the paper is published, and it will not have a break change.

Improvement:

- The overfitting issue is solved by using smaller batch_size on rmd17;
- The NaN problem is alleviated by using a new concatenate-splitting node.

### CODE REVIEW

File: tests/test_eqvar.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            energy1 = pinet(batch1)
+
+            batch2 = batch.copy()
+            batch2['coord'] = rotate(tf.cast(batch2['coord'], tf.float32), random.randint(0, 4123))
+            energy2 = pinet(batch2)
+            np.testing.assert_allclose(energy1, energy2, rtol=1e-4, atol=1e-4)
+
+
+    def test_pinet2_p5_irrep_prod_internal(self, mocked_data):
+        from pinn.layers import (
```
Comment (yqshao) [inline tests/test_eqvar.py:116]: Is this meant to test the p3 variable after the whole PiNet? I don't think it's necessary to replicate the whole code here. `pinet2_p5_prod(tensors); tensors['p3']` should do.

File: tests/conftest.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import os
+os.environ["CUDA_VISIBLE_DEVICES"] = "-1"
```
Comment (yqshao) [inline tests/conftest.py:2]: not sure why this is needed here (this module does not seem to use GPU)

Comment (Roy-Kid) [reply tests/conftest.py:2]: OK, I will remove it. I think it better to run the tests on CPU instead of GPU, especially on HPC's login node(it will use all the cards).

File: pinn/networks/__init__.py [hunk 3]
```diff
@@ -11,11 +11,17 @@ def get(network_spec):
...
+    from pinn.networks.pinet2_p5_combine import PiNet2P5Combine
     implemented_networks = {
         'PiNet': PiNet,
         'BPNN': BPNN,
         'LJ': LJ,
-        'PiNet2': PiNet2
+        'PiNet2': PiNet2,
+        'PiNet2P5Dot': PiNet2P5Dot,
+        'PiNet2P5Prod': PiNet2P5Prod,
+        'PiNet2P5Combine': PiNet2P5Combine,
```
[the remark this answers is not in the document]
Comment (Roy-Kid) [reply pinn/networks/__init__.py:24]: After refactoring pinet2, it should backward support PiNet2-p3 and PiNet2-p5. Also, redundant styles such as dot/prod/combine will be removed, only pinet2 is remained.

File: pinn/networks/pinet2.py [hunk 4]
```diff
@@ -183,7 +183,7 @@ def call(self, tensors):
         Returns:
             output (tensor): an updated output tensor with shape `(n_atoms, out_units)`
         """
-        ind_1, p1, p3, prev_output = tensors
```
Comment (LKnijff) [inline pinn/networks/pinet2.py:186]: Why did you remove p3 here?

Comment (Roy-Kid) [reply pinn/networks/pinet2.py:186]: To reuse outlayer for other variables, for example, i3, i5, p5 etc. ?

Comment (LKnijff) [reply pinn/networks/pinet2.py:186]: Then should p1 be renamed to "prop" or something like in the documentation? To indicate general use?

File: pinn/networks/pinet2_modularized.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        basis_type="polynomial",
+        n_basis=4,
+        gamma=3.0,
+        center=None,
+        pp_nodes=[16, 16],
+        pi_nodes=[16, 16],
+        ii_nodes=[16, 16],
+        out_nodes=[16, 16],
+        out_units=1,
+        out_extra={},
```
Comment (LKnijff) [inline pinn/networks/pinet2_modularized.py:170]: "out_extra" and how to use it is not in the documentation of this function. The same goes for "out_units".

File: nextflow/restart.nf [hunk 6]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env nextflow
+
```
[the remark this answers is not in the document]
Comment (Roy-Kid) [reply nextflow/restart.nf:2]: Ok, I will remove it. But this restart script based on the nextflow is better than a standalone script. 

1. is not relevant to the restart topic because this script only cares about workdir, which is the intermediate fine; 
2. We can not afford to rerun since each training takes more than one day. It also needs to find out which tasks failed and restart them separately(which here must use a script);
3. no matter what language I used here(Groovy / script), it still need to run by using nextflow. After restarting, it can link final result to publicDir. 
4. Since I don't find an easy way to group tasks and runs in a same workdir, it is better to use different workdir for different experiment...

File: tests/test_eqvar.py (hunk 1 above)
Comment (Roy-Kid) [reply tests/test_eqvar.py:116]: Can't. `Preprocessing` copies the tensor, and layers won't change input tensors; I can't find a way to expose intermediate variables from the network.

File: pinn/networks/pinet2_p5_combine.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        tensors = tensors.copy()
+        for k in ["elems", "dist"]:
+            if k in tensors.keys():
+                tensors[k] = tf.reshape(tensors[k], tf.shape(tensors[k])[:1])
+        if "ind_2" not in tensors:
+            tensors.update(self.nl_layer(tensors))
+            tensors["p1"] = tf.cast(  # difference with pinet: prop->p1
+                self.embed(tensors["elems"]), tensors["coord"].dtype
+            )
+        tensors["norm_diff"] = tensors["diff"] / tf.linalg.norm(tensors["diff"])
```
[the remark this answers is not in the document]
Comment (Roy-Kid) [reply pinn/networks/pinet2_p5_combine.py:154]: Shell we keep the `PiNet2`? under this architecture, I think we can only keep one `PiNet2`, which can support rank 1 3 5?

[the remark this answers is not in the document]
Comment (Roy-Kid) [reply pinn/networks/pinet2_p5_combine.py:154]: little bit hard to implement manual back-prop... move it out first....

File: nextflow/restart.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (yqshao) [reply nextflow/restart.nf:2]: Ok, consider this resolved then. For the sake of restarting, I think NF logs the tasks for each run, see [`nextflow log [runname | sessionid]`](https://www.nextflow.io/docs/latest/cli.html#log), that's what I mean by "workdirs of that run".

File: tests/test_eqvar.py (hunk 1 above)
Comment (yqshao) [reply tests/test_eqvar.py:116]: Ah, right, that was to prevent layers from mutating, maybe give the network a `_tensors` attribute to access the intemediate dict of tensors?

Comment (Roy-Kid) [reply tests/test_eqvar.py:116]: Lisanne wants to return internal variables and transform them into desired channels. I try to return an extra variable, but it looks super ugly https://github.com/Teoroo-CMC/PiNN/pull/18#discussion_r1644197428.

File: pinn/networks/pinet2_splitting.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+            output (tensor): output tensor with shape `[n_atoms, out_nodes]`
+        """
+        tensors = self.preprocess(tensors)
+        tensors["p3"] = tf.zeros([tf.shape(tensors["ind_1"])[0], 3, 1])
+        fc = self.cutoff(tensors["dist"])
+        basis = self.basis_fn(tensors["dist"], fc=fc)
+        output = 0.0
+        for i in range(self.depth):
+            p1, p3 = self.gc_blocks[i](
+                [tensors["ind_2"], tensors["p1"], tensors["p3"], tensors["diff"], basis]
```
Comment (zzy2014) [inline pinn/networks/pinet2_splitting.py:349]: Here we should replace the tensors["diff"] to tensors["norm_diff"]

Comment (Roy-Kid) [reply pinn/networks/pinet2_splitting.py:349]: those architecture will be remove except pinet2 & pinet2_modularized
